### ISSUE TITLE
Badly written loop in Modelica.Electrical.Analog.Basic.M_Transformer

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/Modelica#1319]
**Reported by stefanv on 10 Oct 2013 18:28 UTC**
The **algorithm** section in `Modelica.Electrical.Analog.Basic.M_Transformer` looks like this:

```mo
  for s in 1:N loop
     for z in 1:N loop
       Lm[z,s]:= if (z>=s) then L[(s-1)*N+z-div((s-1)*s,2)] else
                 Lm[s,z];
     end for;
  end for;
```

Notice that the value of `Lm[z,s]` is assigned either some expression involving `L` (not `Lm`), or some other entry of `Lm`.

This makes the array `Lm` both an input and an output of the algorithm, meaning that any correct code that a tool generates for the algorithm must first initialize `Lm` to something, namely `Lm.start`.

The code also assumes that a **for** loop iterates over its vector expression sequentially, from first to last entry. There is nothing in the specification that states that this is the case, and I don't think that that is intended to be the case.

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/1319

**Comment by modelica-trac-importer** [comment]
**Comment by stefanv on 10 Oct 2013 18:57 UTC**
So as to provide solutions, and not just complaints, I think the code should have been written like this:

```mo
protected
  Real t;
algorithm
  for s in 1:N loop
     for z in s:N loop
       t := L[(s-1)*N+z-div((s-1)*s,2)];
       Lm[s,z] := t;
       Lm[z,s] := t;
     end for;
  end for;
```

**Comment by modelica-trac-importer** [comment]
**Comment by anonymous on 10 Oct 2013 21:08 UTC**
```mo
protected
  Integer t;
algorithm
  for s in 1:N loop
    t := (s-1)*N-div((s-1)*s,2);
    for z in s:N loop
      Lm[s,z] := L[t+z];
      Lm[z,s] := L[t+z];
    end for;
  end for;
```

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 8 May 2014 09:02 UTC**
In 7a56919749696922fbc42ea8ad17737e222f469b for #1045 the algorithm section was rewritten as inital equation section, but still contains the badly written loop. This is a new proposal:

```mo
protected
  parameter Integer tmp[N](each final fixed=false);
initial equation
  for s in 1:N loop
    tmp[s] = (s-1)*N-div((s-1)*s,2);
    for z in s:N loop
      Lm[s,z] = L[tmp[s]+z];
      Lm[z,s] = L[tmp[s]+z];
    end for;
  end for;
```

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 29 Oct 2015 21:41 UTC**
Resolved by 8a948d81182a362a7d8089fee6a8a3b04cbc7ccc. No auxiliary variable was introduced at the price of some redundant code and computational effort.

**Comment by modelica-trac-importer** [comment]
**Changelog modified by beutlich on 29 Oct 2015 21:41 UTC**
Rewrote loop for initial computation of symmetric matrix Modelica.Electrical.Analog.Basic.M_Transformer.Lm
